Classes/Capture/AddNoteViewController.swift

```diff
@@ -54,12 +54,114 @@ final class AddNoteViewController: UIViewController {
+        guard self.textView.text != nil else { return } // If no selected text, ignore.  Consider popup?
+        if let range = self.textView.selectedTextRange, !range.isEmpty {
+            let selectedText = self.textView.text( in: range )
+            self.textView.replace( range, withText: "[[\(selectedText!)][]]" )
+            let newCursor = self.textView.position(from: range.end, offset: 4)!
+            self.textView.selectedTextRange = self.textView.textRange( from: newCursor, to: newCursor)
+        }
+    }
+
+    let datePicker: UIDatePicker = {
+        let picker = UIDatePicker(frame: CGRect(x: 0, y: 0, width: UIScreen.main.bounds.width, height: 216))
+        picker.datePickerMode = .dateAndTime
+        return picker
+    }()
+
+    private lazy var datePickBar: UIToolbar = {
+        let bar = UIToolbar()
+        let cancelButton = UIBarButtonItem( title: "Cancel", style: .plain, target: self, action: #selector(setDefaultInputView) )
+        let scheduleButton = UIBarButtonItem( title: "Schd", style: .plain, target: self, action: #selector(insertDateSchedule) )
+        let deadlineButton = UIBarButtonItem( title: "Dead", style: .plain, target: self, action: #selector(insertDateDeadline) )
+        let agendaButton = UIBarButtonItem( title: "Agenda", style: .plain, target: self, action: #selector(insertDateAgenda) )
+        let plainButton = UIBarButtonItem( title: "Plain", style: .plain, target: self, action: #selector(insertDatePlain) )
+        let spacer = UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
+        bar.items = [ scheduleButton, deadlineButton, agendaButton, plainButton, spacer, cancelButton ]
+        bar.sizeToFit()
+        return bar
+    }()
+
+    private func datePicked() -> String {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "YYYY-MM-dd E hh:mm"
+        //dateFormatter.formatOptions = [.withTime]
+        return dateFormatter.string( from: self.datePicker.date )
+    }
+
+    private func replaceSelected( with newString: String ) {
+        print( "replaceSelected" )
+        if self.textView.text != nil {
+            if let range = self.textView.selectedTextRange {
+                self.textView.replace( range, withText: newString )
+            }
+        }
+    }
+
+    @objc func insertDateSchedule() {
+        let date = datePicked()
+        replaceSelected( with: "SCHEDULED: <\(date)>" )
+        self.setDefaultInputView()
+    }
+
+    @objc func insertDateDeadline() {
+        let date = datePicked()
+        replaceSelected( with: "DEADLINE: <\(date)>" )
+        self.setDefaultInputView()
+    }
+
+    @objc func insertDateAgenda() {
+        let date = datePicked()
+        replaceSelected( with: "<\(date)>" )
+        self.setDefaultInputView()
+    }
+
+    @objc func insertDatePlain() {
+        let date = datePicked()
+        replaceSelected( with: "[\(date)]" )
+        self.setDefaultInputView()
+    }
```

Comment:
This one is a huge proposal :) I do not like much all these functions that do almost the same. We can improve a bit with the following proposals.

Create a `enum` to store text for our labels:
```swift
enum DateType: String {
    case schedule = "Schd"
    case deadline = "Dead"
    case agenda = "Agenda"
    case plain = "Plain"
}
``

Create a function that will be able to distinguish these types based on the string value:
```swift
@objc private func insertDate(_ sender: UIBarButtonItem) {
    guard let title = sender.title else { return }
    let date = datePicked()
    switch DateType(rawValue: title) {
    case .agenda:
        replaceSelected( with: "<\(date)>" )
    case .deadline:
        replaceSelected( with: "DEADLINE: <\(date)>" )
    case .plain:
        replaceSelected( with: "[\(date)]" )
    case .schedule:
        replaceSelected( with: "SCHEDULED: <\(date)>" )
    case .none:
        break
    }
    self.setDefaultInputView()
}
```

Rewrite `datePickBar` lazy variable function to pass the new `insertDate(:)` to all `UIBarButtonItem` we create:
```swift
private lazy var datePickBar: UIToolbar = {
    let bar = UIToolbar()
    let cancelButton = UIBarButtonItem( title: "Cancel", style: .plain, target: self, action: #selector(setDefaultInputView) )
    let scheduleButton = UIBarButtonItem( title: DateType.schedule.rawValue, style: .plain, target: self, action: #selector(insertDate(_:)) )
    let deadlineButton = UIBarButtonItem( title: DateType.deadline.rawValue, style: .plain, target: self, action: #selector(insertDate(_:)) )
    let agendaButton = UIBarButtonItem( title: DateType.agenda.rawValue, style: .plain, target: self, action: #selector(insertDate(_:)) )
    let plainButton = UIBarButtonItem( title: DateType.plain.rawValue, style: .plain, target: self, action: #selector(insertDate(_:)) )
    let spacer = UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
    bar.items = [ scheduleButton, deadlineButton, agendaButton, plainButton, spacer, cancelButton ]
    bar.sizeToFit()
    return bar
}()
```

Delete the functions marked above.

In general, this approach is more or less universal. It stores text in the `enum`, will control new cases in the `enum` and notify that we have to handle them, etc.

Reply:
I agree this is nicer.  I didn't know about passing the sender in the selector and so wasn't clear how to call a parameterized function.

Reply:
done.  Thanks again.